File: src/ax_eval/runner.py
"""The runner: orchestrate N tasks x M harnesses into a result set + matrix."""

from __future__ import annotations

from dataclasses import dataclass, field

from .datatypes import RunResult, TargetPack
from .harnesses import get_harness


@dataclass
class RunReport:
    """The full output of a run: every RunResult plus the pack/harness context."""

    pack: str
    pack_version: str
    harnesses: list[str]
    results: list[RunResult] = field(default_factory=list)

    def matrix(self) -> dict[str, dict[str, bool]]:
        """task_id -> {harness -> success}."""
        grid: dict[str, dict[str, bool]] = {}
        for r in self.results:
            grid.setdefault(r.task_id, {})[r.harness] = r.success
        return grid

    def pass_rate(self, harness: str) -> float:
        rs = [r for r in self.results if r.harness == harness]
        return sum(r.success for r in rs) / len(rs) if rs else 0.0


def run(
    pack: TargetPack,
    harness_names: list[str],
    *,
    progress: bool = False,
) -> RunReport:
    """Run every task in ``pack`` across each named harness, sequentially."""
    report = RunReport(
        pack=pack.name, pack_version=pack.version, harnesses=list(harness_names)
    )
    for name in harness_names:
        harness = get_harness(name)
        for task in pack.tasks:
            result = harness.run(task, pack)
            report.results.append(result)
            if progress:
                mark = "PASS" if result.success else "FAIL"
                print(f"  [{mark}] {name} x {task.id}")
    return report

File: src/runner.ts
/** The runner: orchestrate N tasks × M harnesses into a result set + matrix. */
import { getHarness } from "./adapters/registry.js";
import type { RunResult, TargetPack } from "./schemas.js";

export interface RunReport {
  pack: string;
  packVersion: string;
  harnesses: string[];
  /** Subset of `harnesses` that are synthetic controls (e.g. a perfect mock).
   *  Excluded from the static×behavioral gap so a by-construction ceiling can't
   *  erase it. */
  synthetic: string[];
  results: RunResult[];
}

export interface RunOptions {
  progress?: boolean;
}

/** Run every task in `pack` across each named harness, sequentially. */
export async function run(
  pack: TargetPack,
  harnessNames: string[],
  opts: RunOptions = {},
): Promise<RunReport> {
  const results: RunResult[] = [];
  const synthetic: string[] = [];
  for (const name of harnessNames) {
    const harness = getHarness(name);
    if (harness.synthetic) synthetic.push(name);
    for (const task of pack.tasks) {
      const result = await harness.run(task, pack);
      results.push(result);
      if (opts.progress) {
        const mark = result.success ? "PASS" : "FAIL";
        console.log(`  [${mark}] ${name} × ${task.id}`);
      }
    }
  }
  return { pack: pack.name, packVersion: pack.version, harnesses: [...harnessNames], synthetic, results };
}

/** task_id → { harness → success }. */
export function matrix(report: RunReport): Record<string, Record<string, boolean>> {
  const grid: Record<string, Record<string, boolean>> = {};
  for (const r of report.results) {
    (grid[r.taskId] ??= {})[r.harness] = r.success;
  }
  return grid;
}

export function passRate(report: RunReport, harness: string): number {
  const rs = report.results.filter((r) => r.harness === harness);
  return rs.length ? rs.filter((r) => r.success).length / rs.length : 0;
}
